### PR TITLE
[bridge/ot] Fall-back to TextMap carrier when it's not ot.HttpHeaders

### LINKED ISSUE #3678: [opentracing] OT Bridge does not work with OT gRPC instrumentation
### Description

The opentracing bridge is not able to inject/extract trace context in gRPC interceptors. The failure is in this code (and in similar place in `extract`): https://github.com/open-telemetry/opentelemetry-go/blob/c9cb53c5b416c116dff55c30a31b3e9f7901cdaf/bridge/opentracing/bridge.go#L667-L670

Here, the OT carrier is required to be a helper type `ot.HTTPHeadersCarrier`, which is not a requirement of the OpenTracing API. In contrast, here is the "official" gRPC-OT instrumentation code that passes a private adapter type as a carrier: https://github.com/grpc-ecosystem/grpc-opentracing/blob/8e809c8a86450a29b90dcc9efbf062d0fe6d9746/go/otgrpc/client.go#L232

This is unnecessary restriction since a general OT interface can still be used (with adapter) to satisfy `propagation.*` carrier interface, as is already done with `textMapWrapper`.

### Environment

- opentelemetry-go version: v1.11.2

### Other questions

There is a valid question: why existing gRPC instrumentation uses `ot.HttpHeaders` carrier type instead of `ot.TextMap`? I tried changing it to use TextMap type and the propagation seems to work (using W3C header format, without tracestate). However, this change might result in differences in how the header values are encoded (TextMap is supposed to be as-it, HttpHeaders are supposed to be url-encoded), and could break some installations. So it's better to fix it here since the current bridge is being overly restrictive in the type of carrier it allows, which does not match OpenTracing API expectations.

### PR DESCRIPTION
**Pull request by yurishkuro**
Resolves #3678

### CODE REVIEW

**Review by MadVikingGod — COMMENTED** [review]
It looks like this assumes that if the format is `HTTPHeaders` and we fall back to the TextMapCarrier, the backing type will do the appropriate URL encoding.   I would suggest we document this edge case.

File: .gitignore [hunk 1]
```diff
@@ -8,6 +8,9 @@ Thumbs.db
 *.so
 coverage.*
 
+go.work
+go.work.sum
+
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply .gitignore:13]: Are you saying .gitignore needs a whole separate PR?

**Comment by MadVikingGod** [comment]
`http.Header.Set()` does canonicalize keys, but it does look like it leaves values alone.

**Review by MadVikingGod — APPROVED** [review]

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (MadVikingGod) [reply .gitignore:13]: Yes making changes around go workspaces deserves a bit more research than just including it in an unrelated PR.

[the remark this answers is not in the document]
Comment (yurishkuro) [reply .gitignore]: https://github.com/open-telemetry/opentelemetry-go/pull/3681

**Comment by yurishkuro** [comment]
can this be merged?

File: bridge/opentracing/go.mod [hunk 2]
```diff
@@ -5,17 +5,25 @@ go 1.18
 replace go.opentelemetry.io/otel => ../..
 
 require (
+	github.com/opentracing-contrib/go-grpc v0.0.0-20210225150812-73cb765af46e
 	github.com/opentracing/opentracing-go v1.2.0
 	github.com/stretchr/testify v1.8.1
 	go.opentelemetry.io/otel v1.13.0
 	go.opentelemetry.io/otel/trace v1.13.0
+	google.golang.org/grpc v1.52.3
```
Comment (Aneurysm9) [inline bridge/opentracing/go.mod:13]: Are these dependencies pulled in only for testing?  If so, can they be moved to a `bridge/opentracing/test` module to minimize the impact on dependents that do not use gRPC and protobuf?

Comment (yurishkuro) [reply bridge/opentracing/go.mod:13]: good idea. But what do you do about code coverage in this case, is the CI setup to count usage from other modules as coverage?

Comment (MrAlias) [reply bridge/opentracing/go.mod:13]: > good idea. But what do you do about code coverage in this case, is the CI setup to count usage from other modules as coverage?

Yes, it is setup to do that. 👍

**Review by Aneurysm9 — APPROVED** [review]

**Pull request merged**